src/core/public/overlays/flyout.test.tsx

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { mockReactDomRender, mockReactDomUnmount } from './flyout.test.mocks';
+
+import React from 'react';
+import { i18nServiceMock } from '../i18n/i18n_service.mock';
+import { FlyoutRef, FlyoutService } from './flyout';
+
+const i18nMock = i18nServiceMock.createSetupContract();
+
+beforeEach(() => {
+  mockReactDomRender.mockClear();
+  mockReactDomUnmount.mockClear();
+});
+
+describe('FlyoutService', () => {
+  describe('openFlyout()', () => {
+    it('renders a flyout to the DOM', () => {
+      const target = document.createElement('div');
+      const flyoutService = new FlyoutService(target);
+      expect(mockReactDomRender).not.toHaveBeenCalled();
+      flyoutService.openFlyout(i18nMock, <span>Flyout content</span>);
+      expect(mockReactDomRender.mock.calls).toMatchSnapshot();
+    });
+    describe('with a currently active flyout', () => {
+      // tslint:disable-next-line: one-variable-per-declaration
+      let target: HTMLElement, flyoutService: FlyoutService, ref1: FlyoutRef;
+      beforeEach(() => {
+        target = document.createElement('div');
+        flyoutService = new FlyoutService(target);
+        ref1 = flyoutService.openFlyout(i18nMock, <span>Flyout content 1</span>);
+      });
+      it('replaces the current flyout with a new one', () => {
+        flyoutService.openFlyout(i18nMock, <span>Flyout content 2</span>);
+        expect(mockReactDomRender.mock.calls).toMatchSnapshot();
+        expect(mockReactDomUnmount).toHaveBeenCalledTimes(1);
```

Comment:
can we test against a result (DOM tree) ? this makes tests less implementation-aware

Reply:
I agree, this also made it really hard to write a good test for instance for rendering into different target DOM elements from the same overlay service (now deprecated functionality). I think `React.render` is used in a non-standard way. Normally there's one call to `React.render` to mount your component tree after which React takes care of your DOM. We instead continually mutate the DOM by re-rendering new react component trees, instead of letting React mutate a single React component tree.

For this reason I couldn't get standard enzyme tests which mount a component to work. I had some success using 'react-test-renderer/shallow' instead of a jest mock for `React.render`. But started feeling if the code is so hard to test, maybe the architecture is wrong. I'd like to explore using a React portal for some core services like the flyout and toasts to see if these could be more easily tested/used. If it doesn't work I'll hook up 'react-test-renderer/shallow' here and in other tests where we use plain Jest mocks for `React.render`.